storage/redis.go

```diff
@@ -57,53 +58,194 @@ func (r *redisStore) CheckRtt(ctx context.Context) int64 {
+	heartbeatJson, err := json.Marshal(heartbeatStrMap)
+	if err != nil {
+		return nil, err
+	}
 
-	pipe.HSet(ctx, managerIdentity, "beat", heartbeat.Beat.UTC().Unix())
-	pipe.HSet(ctx, managerIdentity, "quiet", heartbeat.Quiet)
-	pipe.HSet(ctx, managerIdentity, "busy", heartbeat.Busy)
-	pipe.HSet(ctx, managerIdentity, "rtt_us", rtt)
-	pipe.HSet(ctx, managerIdentity, "rss", heartbeat.RSS)
-	pipe.HSet(ctx, managerIdentity, "info", heartbeat.Info)
-	pipe.Expire(ctx, managerIdentity, 60*time.Second) // set the TTL of the heartbeat to 60
+	heartbeat := Heartbeat{}
+	err = json.Unmarshal(heartbeatJson, &heartbeat)
+	if err != nil {
+		return nil, err
+	}
+	heartbeat.Identity = heartbeatID
+	return &heartbeat, nil
+}
 
-	workersKey := managerIdentity + sidekiqHeartbeatJobsKey
+func (r *redisStore) GetAllHeartbeats(ctx context.Context) ([]*Heartbeat, error) {
+	var heartbeats []*Heartbeat
 
-	pipe.Del(ctx, workersKey)
+	heartbeatIDs, err := r.getHeartbeatIDs(ctx)
+	if len(heartbeatIDs) == 0 {
+		return nil, err
+	}
+	for _, heartbeatID := range heartbeatIDs {
+		heartbeat, err := r.getHeartbeat(ctx, heartbeatID)
+		if err != nil {
+			return nil, err
+		}
+		if heartbeat != nil {
+			heartbeats = append(heartbeats, heartbeat)
+		}
+	}
+	return heartbeats, nil
+}
 
-	for tid, msg := range heartbeat.WorkerMessages {
-		// fake the sidekiq thread id
-		fakeThreadId := fmt.Sprintf("%d-%s", heartbeat.Pid, tid)
-		pipe.HSet(ctx, workersKey, fakeThreadId, msg)
+func (r *redisStore) getHeartbeatIDs(ctx context.Context) ([]string, error) {
+	heartbeatIDs, err := r.client.SMembers(ctx, GetProcessesKey(r.namespace)).Result()
+	if err != nil && err != redis.Nil {
+		return nil, err
 	}
+	return heartbeatIDs, nil
+}
 
-	pipe.Expire(ctx, workersKey, 60*time.Second)
+func (r *redisStore) SendHeartbeat(ctx context.Context, heartbeat *Heartbeat) error {
+	pipe := r.client.Pipeline()
+	rtt := r.CheckRtt(ctx)
 
-	_, err := pipe.Exec(ctx)
+	managerKey := GetManagerKey(r.namespace, heartbeat.Identity)
+	pipe.SAdd(ctx, GetProcessesKey(r.namespace), heartbeat.Identity) // add to the sidekiq processes set without the namespace
+
+	workerHeartbeats, err := json.Marshal(heartbeat.WorkerHeartbeats)
+	if err != nil {
+		return err
+	}
+
+	pipe.HMSet(ctx, managerKey,
+		"beat", heartbeat.Beat,
+		"quiet", heartbeat.Quiet,
+		"busy", heartbeat.Busy,
+		"rtt_us", rtt,
+		"rss", heartbeat.RSS,
+		"info", heartbeat.Info,
+		"manager_priority", heartbeat.ManagerPriority,
+		"active_manager", heartbeat.ActiveManager,
+		"worker_heartbeats", workerHeartbeats)
+
+	_, err = pipe.Exec(ctx)
 	if err != nil && err != redis.Nil {
 		return err
 	}
 
 	return nil
 }
 
-func (r *redisStore) RemoveHeartbeat(ctx context.Context, heartbeat *Heartbeat) error {
-	managerIdentity := r.namespace + heartbeat.Identity
+func (r *redisStore) HandleAllExpiredHeartbeats(ctx context.Context, expireTS int64) ([]*StaleMessageUpdate, error) {
```

Comment:
Can this code be moved into the manager? And use a combination of `GetAllHeartbeats` and `RequeueMessagesFromInProgressQueue`?

Reply:
Moved into manager in 5dcbe1fc2191115664209dc63ea3d5a228911580.

Reply:
Awesome work!